domains/certificates/Query.API/API.AppTests/ContractTests.cs

```diff
@@ -205,5 +242,141 @@ public async Task GetSpecificContract_UserIsNotOwner_NotFound()
+
+        response.StatusCode.Should().Be(HttpStatusCode.Created);
+
+        var endContractBody = new
+        {
+            gsrn,
+            endDate = DateTimeOffset.Now.AddDays(3).ToUnixTimeSeconds()
+        };
+
+        var endContractResponse = await client.PatchAsJsonAsync("api/certificates/contracts", endContractBody);
+
+        endContractResponse.StatusCode.Should().Be(HttpStatusCode.OK);
+    }
+
+    [Fact]
+    public async Task EndContract_WithoutEndDate_Ended()
+    {
+        var gsrn = GsrnHelper.GenerateRandom();
+        dataSyncWireMock.SetupMeteringPointsResponse(gsrn);
+
+        var subject = Guid.NewGuid().ToString();
+        using var client = factory.CreateAuthenticatedClient(subject);
+
+        var body = new
+        {
+            gsrn,
+            startDate = DateTimeOffset.Now.ToUnixTimeSeconds()
+        };
+
+        var response = await client.PostAsJsonAsync("api/certificates/contracts", body);
+
+        response.StatusCode.Should().Be(HttpStatusCode.Created);
+
+        var endContractBody = new
+        {
+            gsrn
+        };
+
+        var endContractResponse = await client.PatchAsJsonAsync("api/certificates/contracts", endContractBody);
+
+        endContractResponse.StatusCode.Should().Be(HttpStatusCode.OK);
+    }
+
+    [Fact]
+    public async Task EndContract_NoContractCreated_NoContract()
+    {
+        var gsrn = GsrnHelper.GenerateRandom();
+        dataSyncWireMock.SetupMeteringPointsResponse(gsrn);
+
+        var subject = Guid.NewGuid().ToString();
+        using var client = factory.CreateAuthenticatedClient(subject);
+
+        var body = new
+        {
+            gsrn,
+            endDate = DateTimeOffset.Now.AddDays(3).ToUnixTimeSeconds()
+        };
+
+        var response = await client.PatchAsJsonAsync("api/certificates/contracts", body);
+
+        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
+    }
+
+    [Fact]
+    public async Task EndContract_TimeBefore_BadRequest()
+    {
+        var gsrn = GsrnHelper.GenerateRandom();
+        dataSyncWireMock.SetupMeteringPointsResponse(gsrn);
+
+        var subject = Guid.NewGuid().ToString();
+        using var client = factory.CreateAuthenticatedClient(subject);
+
+        var body = new
+        {
+            gsrn,
+            startDate = DateTimeOffset.Now.ToUnixTimeSeconds()
+        };
+
+        var response = await client.PostAsJsonAsync("api/certificates/contracts", body);
+
+        response.StatusCode.Should().Be(HttpStatusCode.Created);
+
+        dataSyncWireMock.Dispose();
+        dataSyncWireMock.SetupMeteringPointsResponse(gsrn);
+        var endContractBody = new
+        {
+            gsrn,
+            endDate = DateTimeOffset.Now.AddDays(-3).ToUnixTimeSeconds()
+        };
+
+        var endContractResponse = await client.PatchAsJsonAsync("api/certificates/contracts", endContractBody);
+
+        endContractResponse.StatusCode.Should().Be(HttpStatusCode.BadRequest);
+    }
+
```

Comment:
There is no test case for when a contract exists and a user that is not the owner, tries to change it.

Reply:
Thats due to that i couldn't figure out a way to generate or choose an different owner

Reply:
You can create a new client with `using var client = factory.CreateAuthenticatedClient(subject);` If you do this with two different subjects, then you have the case with two different users/clients